versioned_docs/version-4.0.0-alpha.1/datastructures/mapping.md

```diff
@@ -3,76 +3,121 @@ title: Working with Mapping
+        /// Credit more money to the contract.
+        #[ink(message, payable)]
+        pub fn transfer(&mut self) {
+            let caller = self.env().caller();
+            let balance = self.balances.get(caller).unwrap_or(0);
+            let endowment = self.env().transferred_value();
+            self.balances.insert(caller, &(balance + endowment));
+        }
+
+        /// Withdraw all your balance from the contract.
+        pub fn withdraw(&mut self) {
+            let caller = self.env().caller();
+            let balance = self.balances.get(caller).unwrap();
+            self.balances.remove(caller);
+            self.env().transfer(caller, balance).unwrap()
         }
     }
 }
+
 ```
+
+## Considerations when using the `Mapping` type
+
+### Updating values
+
+The attentive reader may have noticed that accessing mapping values via the `Mapping::get()` 
+method will result in an owned value (a local copy), as opposed to a direct reference 
+into the storage. Changes to this value won't be reflected in the contracts storage 
+"automatically". To avoid this common pitfall, the value must be inserted again at the same 
+key after it was modified. The `transfer` function from above example illustrates this:
+
+```rust
+pub fn transfer(&mut self) {
+    let caller = self.env().caller();
+    // `balance` is a local value and not a reference to the value on storage!
+    let balance = self.balances.get(caller).unwrap_or(0);
+    let endowment = self.env().transferred_value();
+    // The following line of code would have no effect to the balance of the 
+    // caller stored in contract storage:
+    //
+    // balance += endowment;
+    //
+    // Instead, we use the `insert` function to write it back like so:
+    self.balances.insert(caller, &(balance + endowment));
+}
+```
+
+### Storage loading behaviour
+
+Each `Mapping` value lives under it's own storage key. Briefly, this means that `Mapping`s are 
+lazily loaded in ink!. In other words, if your message only accesses a single key of a 
+mapping, it will not load the whole mapping but only the value being accessed.
+
+```rust
+// This causes only a single storage access and the decoding of a single "MyValue" struct,
+// no matter ho many elements there are inside the mapping.
+let foo: MyValue = my_mapping.get(0)?;
+
+for n in 0..5 {
+    // This causes a storage access and a decoding operation for each loop iteration.
+    // It is not possible to "fetch" all key/value pairs directly at once.
+    let bar: MyValue = my_mapping.get(n)?;
+}
+```
+
+Furthermore, it follows that mapping values do not have a contiguos storage layout and it is
+therefore not possible to iterate over the contents of a map. 
+Circumventing this restriction by storing populated keys inside an `ink_prelude::Vec` might 
+not always be advisable: As accessing a storage cell is relatively expensive, this might 
+result in very high gas costs for large mappings.
```

Comment:
Also, we don't have `ink_prelude` anymore. You should use `ink::prelude::vec::Vec`

Reply:
> Hmm, maybe it is better not to mention it here or create a separate section to describe how iterable storage could be implemented.

I see the point, but I don't think this is something that fits into the documentation. I just saw fit to mention that as a warning here (it might not be obvious to anyone that loading from storage causes an API call each time and this can get costly).

Reply:
I don't like it here, because it is a big topic. If you use `Vec`, and you have a lot of keys, it will break your contract, it will not fit into the decode buffer. We don't need to propose a solution to the iterator problem.

Also, we plan to support iterating over elements of `Mapping`, because of `blake2_128_concat`, in the future.

I would prefer to keep only the line "you can't iterate over the mapping". Or "you can't iterate over the mapping, but it is possible to implement this behaviour in another way [](link to the article/section/discussion about it)"